### PR TITLE
Add analytics and refactor system design

### PR DESCRIPTION
**Pull request by irenenikk**

### CODE REVIEW

File: src/main/java/fi/helsinki/cs/tmc/cli/Application.java [hunk 1]
```diff
@@ -20,14 +29,7 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.lang.management.ManagementFactory;
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/Application.java:23]: wildcard import

File: src/main/java/fi/helsinki/cs/tmc/cli/analytics/AnalyticsFacade.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import fi.helsinki.cs.tmc.core.domain.Exercise;
+import fi.helsinki.cs.tmc.core.exceptions.UninitializedHolderException;
+import fi.helsinki.cs.tmc.core.holders.TmcSettingsHolder;
+import fi.helsinki.cs.tmc.spyware.EventSendBuffer;
+import fi.helsinki.cs.tmc.spyware.EventStore;
+import fi.helsinki.cs.tmc.spyware.LoggableEvent;
+import fi.helsinki.cs.tmc.spyware.SpywareSettings;
+import org.slf4j.LoggerFactory;
+
+import java.awt.*;
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/analytics/AnalyticsFacade.java]: wildcard import

File: src/main/java/fi/helsinki/cs/tmc/cli/analytics/AnalyticsFacade.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    public void saveAnalytics(Exercise exercise, String command) {
+        if (!settings.isSpywareEnabled()) {
+            return;
+        }
+        LoggableEvent event = LoggableEventCreator.createEvent(exercise, command);
+        saveEvent(event);
+    }
+
+    private void saveEvent(LoggableEvent event) {
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/analytics/AnalyticsFacade.java]: Maybe this should be the last method since it's the only private one?

File: src/main/java/fi/helsinki/cs/tmc/cli/command/LoginCommand.java [hunk 4]
```diff
@@ -51,7 +49,15 @@ public void run(CliContext context, CommandLine args) {
             return;
         }
 
-        if (!ctx.loadBackendWithoutLogin()) {
+        if (this.ctx.checkIsLoggedIn(true, true)) {
+            Optional<String> username = this.ctx.getSettings().getUsername();
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/command/LoginCommand.java:53]: getOrDefault?
same below

File: src/main/java/fi/helsinki/cs/tmc/cli/utils/PropertyFunctions.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package fi.helsinki.cs.tmc.cli.utils;
+
+/**
+ * Created by nikkaire on 5.1.2018.
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/utils/PropertyFunctions.java]: Can you remove this comment?

File: src/test/java/fi/helsinki/cs/tmc/cli/analytics/AnalyticsFacadeTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
+package fi.helsinki.cs.tmc.cli.analytics;
+
+
```
Comment (nygrenh) [inline src/test/java/fi/helsinki/cs/tmc/cli/analytics/AnalyticsFacadeTest.java]: unnecessary whitespace

File: src/test/java/fi/helsinki/cs/tmc/cli/analytics/AnalyticsFacadeTest.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+import fi.helsinki.cs.tmc.cli.Application;
+import fi.helsinki.cs.tmc.cli.backend.*;
```
Comment (nygrenh) [inline src/test/java/fi/helsinki/cs/tmc/cli/analytics/AnalyticsFacadeTest.java]: Wildcard import

File: src/test/java/fi/helsinki/cs/tmc/cli/command/ConfigCommandTest.java [hunk 8]
```diff
@@ -2,40 +2,103 @@
...
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.when;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
 
 import fi.helsinki.cs.tmc.cli.Application;
-import fi.helsinki.cs.tmc.cli.backend.Settings;
+import fi.helsinki.cs.tmc.cli.analytics.AnalyticsFacade;
+import fi.helsinki.cs.tmc.cli.backend.*;
```
Comment (nygrenh) [inline src/test/java/fi/helsinki/cs/tmc/cli/command/ConfigCommandTest.java]: Wildcard import

**Review by nygrenh — CHANGES_REQUESTED** [review]
🐫

File: src/main/java/fi/helsinki/cs/tmc/cli/command/LoginCommand.java (hunk 4 above)
Comment (irenenikk) [reply src/main/java/fi/helsinki/cs/tmc/cli/command/LoginCommand.java:53]: I don't see how getOrDefault would help here since we want to concatenate stuff to the value of the optional

**Review by nygrenh — APPROVED** [review]
